shims/spark310/src/main/scala/com/nvidia/spark/rapids/shims/spark310/ParquetCachedBatchSerializer.scala

```diff
@@ -244,14 +315,124 @@ class ParquetCachedBatchSerializer extends CachedBatchSerializer with Arm {
+
+          override def hasNext: Boolean = (iter != null && iter.hasNext) || cbIter.hasNext
+
+          override def next(): InternalRow = {
+            if (iter == null || !iter.hasNext) {
+              iter = convertColumnarBatchToInternalRowIter()
+            }
+            if (iter.hasNext) {
+              iter.next()
+            } else {
+              InternalRow.empty
+            }
+          }
+
+          def convertColumnarBatchToInternalRowIter(): Iterator[InternalRow] = {
+            val parquetCachedBatch = cbIter.next().asInstanceOf[ParquetCachedBatch]
+            val inputFile = new ByteArrayInputFile(parquetCachedBatch.buffer)
+
+            val catalystColumns =
+              getColumnIndices(cacheAttributes).map { i => "_col" + i }
+
+            val catalystSchema = cacheAttributes.zipWithIndex.map {
+              case (attr, i) => attr.withName(catalystColumns(i))
+            }
+
+            val selectedColumns =
+              getColumnIndices(selectedAttributes).map { i => "_col" + i }
+
+            val requestedSchema = selectedAttributes.zipWithIndex.map {
+              case (attr, i) => attr.withName(selectedColumns(i))
+            }
+
+            val sharedConf = broadcastedHadoopConf.value.value
+            val options = HadoopReadOptions.builder(sharedConf).build()
+            val parquetFileReader = ParquetFileReader.open(inputFile, options)
+            val footerFileMetaData = parquetFileReader.getFileMetaData
+            val parquetSchema = parquetFileReader.getFooter.getFileMetaData.getSchema
+
+            val convertTz = None // as we are using hadoop and parquet mr
+
+            val unsafeRows = new ArrayBuffer[InternalRow]
+            import org.apache.parquet.io.ColumnIOFactory
+            var pages = parquetFileReader.readNextRowGroup()
+            while (pages != null) {
+              val rows = pages.getRowCount
+              val columnIO = new ColumnIOFactory().getColumnIO(parquetSchema)
+              val recordReader =
+                columnIO.getRecordReader(pages, new ParquetRecordMaterializer(parquetSchema,
+                  getSchemaFromAttributeSeq(catalystSchema),
+                  new ParquetToSparkSchemaConverter(sharedConf), convertTz,
+                  LegacyBehaviorPolicy.CORRECTED))
+              for (i <- 0 until rows.toInt) {
+                val row = recordReader.read
+                unsafeRows += row.copy()
+              }
+              pages = parquetFileReader.readNextRowGroup()
+            }
+            parquetFileReader.close()
+            val iter = unsafeRows.iterator
+            val unsafeProjection =
+              GenerateUnsafeProjection.generate(requestedSchema, requestedSchema)
+            iter.map(unsafeProjection)
+          }
+        }
+      }
+    }
+  }
+
+  private def getHadoopConf(requestedSchema: StructType,
+     sqlConf: SQLConf) : Configuration = {
+    val hadoopConf = new Configuration()
```

Comment:
Do we really want a fresh Hadoop configuration here?  Wondering if we should instead base this config on `hadoopConfiguration` in the SparkContext to pick up any Hadoop settings the user explicitly specified.

Reply:
You can correct me if I am wrong but I think we do because this code is writing a parquet file that's completely internal to us, no one else will be dealing with the CachedBatch generated by this code but us. Probably doesn't hurt to get the config from SparkContext and then override the configs like the `LEGACY_PARQUET_REBASE_MODE_IN_WRITE` and `LEGACY_PARQUET_REBASE_MODE_IN_READ`.